Add explicit versions in the branches.json and use them for calculating unreleased versions for BwC tests

jozala opened the pull request:
Gradle task added to modify `branches.json`. Possibility to add, remove and update branches in the file with the `version` field added next to the `branch`.
After the change the file looks like this:
```json
{
    "branches": [
        {
            "branch": "main",
            "version": "9.1.0"
        },
        {
            "branch": "9.0",
            "version": "9.0.3"
        },
        {
            "branch": "8.19",
            "version": "8.19.1"
        }
    ]
}
```
BwC test configuration has been changed to use changed `branches.json` file to calculate unreleased ES versions, instead of combining data from `Version.java` and `branches.json`.

jozala committed:
Initial work for having explicit active branch with version connection

# Conflicts:
#	branches.json

jozala committed:
Basic implementation with single test passing

jozala committed:
Tests for UpdateBranchesJsonTask

jozala committed:
Allow changing branch.json location in GlobalBuildInfoPlugin

By default, it now reads branches.json from main branch of Elasticsearch
 (through GitHub raw link)
. This can be overridden for testing or when there is a problem with
remote call.

jozala committed:
Change branch.json location temporarily to point to the forked repo

Commit:
[CI] Auto commit changes from spotless

jozala committed:
Artificial Gradle projects for BwC tests added

jozala committed:
Tests corrected to be in line with new BwC test naming

jozala committed:
Register the task in the plugin

jozala committed:
Configure JSON formatting for branches.json

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/UpdateBranchesJsonTask.java:
```diff
@@ -0,0 +1,152 @@
+
+    @TaskAction
+    public void executeTask() throws IOException {
+        File branchesFile = new File(Util.locateElasticsearchWorkspace(project.getGradle()), "branches.json");
```
referencing `project` here while executing task is deprecated. In this case I think we should just inject `ProjectLayout` instead of project in the ctor and then use `layout.getSettingsDirectory().file('branches.json').asFile`. 

The `Util.locateElasticsearchWorkspace` is just a utility method only used to resolve files relateive from internal builds within this repository, which we don't need here as we never run this task anywhere else but in the :elasticsearch root build

breskeby replied:
Also we probably want to make this file an outputFile property of the task

breskeby commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/UpdateBranchesJsonTask.java:
```diff
@@ -0,0 +1,152 @@
+    private final ObjectMapper objectMapper;
+    private final BranchesFileParser branchesFileParser;
+
+    @Nullable
```
ARe those really nullable? usually we do not use that annotation. In tasks you usually mark properties that are optional as @Optional

jozala replied:
I've made all three parameters `@Input` and `@Optional`.

jozala committed:
Update manually branches.json to make it in line with current releases

jozala committed:
Apply code review feedback

jozala committed:
Fix the test to make it pass

jozala committed:
Change how version.properties is consumed in GlobalBuildInfoPlugin

This change is needed to make  GlobalBuildInfoPlugin testable.
Previously it was using static VersionProperties which loaded values
once for JVM so it was impossible to test with different values.

jozala committed:
Update branches.json after releases

jozala commented on build-tools-internal/src/integTest/resources/org/elasticsearch/gradle/internal/fake_git/remote/settings.gradle:
```diff
@@ -9,14 +9,14 @@
-include ":distribution:bwc:staged"
-include ":distribution:bwc:staged2"
-include ":distribution:bwc:maintenance"
+include ":distribution:bwc:major1"
+include ":distribution:bwc:major2"
+include ":distribution:bwc:major3"
+include ":distribution:bwc:major4"
+include ":distribution:bwc:minor1"
+include ":distribution:bwc:minor2"
+include ":distribution:bwc:minor3"
+include ":distribution:bwc:minor4"
```
This is simplification of the naming convention.
Name `minorN` means previous minor on the same major where number `N` is a distance from the current.
Name `majorN` means previous major, still actively developed versions, where N is a distance from the current, omitting all the previous minor versions.

jozala commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/BwcVersions.java:
```diff
@@ -127,81 +105,46 @@ public void forPreviousUnreleased(Consumer<UnreleasedVersionInfo> consumer) {
-            Version version = versions.stream()
-                .sorted(Comparator.reverseOrder())
-                .filter(v -> v.toString().startsWith(branch))
-                .findFirst()
-                .orElse(null);
-
-            // If we don't know about this version we can ignore it
-            if (version == null) {
-                continue;
-            }
-
-            // If this is the current version we can ignore as we've already handled it
-            if (version.equals(currentVersion)) {
-                continue;
-            }
-
-            // We only maintain compatibility back one major so ignore anything older
-            if (currentVersion.getMajor() - version.getMajor() > 1) {
-                continue;
-            }
-
-            // This is the maintenance version
-            if (i == featureFreezeBranches.size() - 1) {
-                result.put(version, new UnreleasedVersionInfo(version, branch, ":distribution:bwc:maintenance"));
-            } else if (version.getRevision() == 0) { // This is the next staged minor
-                String project = existingStaged ? "staged2" : "staged";
-                result.put(version, new UnreleasedVersionInfo(version, branch, ":distribution:bwc:" + project));
-                existingStaged = true;
-            } else { // This is a bugfix
-                bugfixCount++;
-                String project = "bugfix" + (bugfixCount > 1 ? bugfixCount : "");
-                result.put(version, new UnreleasedVersionInfo(version, branch, ":distribution:bwc:" + project));
-            }
+        List<DevelopmentBranch> previousMajorBranches = bwcBranches.getOrDefault("major", Collections.emptyList());
+        for (int i = 0; i < previousMajorBranches.size(); i++) {
+            DevelopmentBranch previousMajorBranch = previousMajorBranches.get(i);
+            result.put(
+                previousMajorBranch.version(),
+                new UnreleasedVersionInfo(previousMajorBranch.version(), previousMajorBranch.name(), ":distribution:bwc:major" + (i + 1))
+            );
```
This is the main change. Now we use `branches.json` to get information about the unreleased versions instead of calculating it from `Version.java`.

jozala commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/info/GlobalBuildInfoPlugin.java:
```diff
@@ -189,32 +202,49 @@ private String formatJavaVendorDetails(JvmInstallationMetadata runtimeJdkMetaDat
+    private List<DevelopmentBranch> getDevelopmentBranches() {
+        String branchesFileLocation = project.getProviders()
+            .gradleProperty(BRANCHES_FILE_LOCATION_PROPERTY)
+            .getOrElse(DEFAULT_BRANCHES_FILE_URL);
+        LOGGER.info("Reading branches.json from {}", branchesFileLocation);
+        byte[] branchesBytes;
+        if (branchesFileLocation.startsWith("http")) {
+            try (InputStream in = URI.create(branchesFileLocation).toURL().openStream()) {
+                branchesBytes = in.readAllBytes();
+            } catch (IOException e) {
+                throw new UncheckedIOException("Failed to download branches.json from: " + branchesFileLocation, e);
+            }
+        } else {
+            try {
+                branchesBytes = Files.readAllBytes(new File(branchesFileLocation).toPath());
+            } catch (IOException e) {
+                throw new UncheckedIOException("Failed to read branches.json from: " + branchesFileLocation, e);
             }
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
         }
 
-        return branches;
+        return branchesFileParser.parse(branchesBytes);
```
It reads development branches (with current versions of ES on these branches) from main branch using raw GitHub link. This URL can be overridden with `org.elasticsearch.build.branches-file-location` Gradle property if needed.

jozala committed:
Add version.properties files for the tests

This file is needed now during the tests as it is no longer fully static
 during full JVM execution.

jozala committed:
Merge branch 'main' into bwc-explicit-branches-version

jozala committed:
Remove unused bwc projects

jozala committed:
Update branches.json manually after releases

This will be done automatically when changes in release-tools are merged
 and these changes are on main

brianseeders commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/info/GlobalBuildInfoPlugin.java:
```diff
@@ -51,28 +51,37 @@
 public class GlobalBuildInfoPlugin implements Plugin<Project> {
     private static final Logger LOGGER = Logging.getLogger(GlobalBuildInfoPlugin.class);
     private static final String DEFAULT_VERSION_JAVA_FILE_PATH = "server/src/main/java/org/elasticsearch/Version.java";
+    private static final String DEFAULT_BRANCHES_FILE_URL = "https://raw.githubusercontent.com/elastic/elasticsearch/master/branches.json";
```
I think `master` redirects (for now), but this should be `main`

jozala replied:
Good catch. Fixed.

jozala committed:
Change "master" to "main" branch name in the default URL for branches
.json file

jozala committed:
Update branches.json manually after releases

This will be done automatically when changes in release-tools are merged
 and these changes are on main

jozala committed:
Merge branch 'main' into bwc-explicit-branches-version

jozala committed:
Merge branch 'main' into bwc-explicit-branches-version

jozala committed:
Update branches.json manually after releases

This will be done automatically when changes in release-tools are merged
 and these changes are on main

Commit:
[CI] Auto commit changes from spotless

jozala committed:
Revert the deletion of `setupLocalGitRepo` method

It has been unused earlier, but now used again in
`AbstractTransportVersionFuncTest`

jozala committed:
Merge branch 'main' into bwc-explicit-branches-version

jozala committed:
Merge branch 'main' into bwc-explicit-branches-version

jozala committed:
Fix the compilation issues after merge

Commit:
[CI] Auto commit changes from spotless

jozala committed:
Update branches.json manually after releases

This will be done automatically when changes in release-tools are merged
and these changes are on main

jozala committed:
Merge branch 'main' into bwc-explicit-branches-version

jozala committed:
Remove temporary branches-file-location property in gradle.properties

This change cleans up the configuration by removing the property used for testing changes in BwC before the version field was introduced in branches.json. The associated TODO comment has also been removed.
